File: 21/main.dart
import 'dart:io';
import 'dart:typed_data';
import 'dart:math';

const double EPS = 1e-13;

void main() {
  final builder = BytesBuilder(copy: false);
  stdin.listen((event) {
    builder.add(event);
  }, onDone: () {
    if (builder.isNotEmpty) {
      solve(Scanner(builder.takeBytes()));
    }
  });
}

class Scanner {
  final Uint8List _bytes;
  int _ptr = 0;

  Scanner(this._bytes);

  bool hasNext() {
    while (_ptr < _bytes.length && _bytes[_ptr] <= 32) {
      _ptr++;
    }
    return _ptr < _bytes.length;
  }

  int readInt() {
    while (_ptr < _bytes.length && _bytes[_ptr] <= 32) {
      _ptr++;
    }
    int sign = 1;
    if (_ptr < _bytes.length && _bytes[_ptr] == 45) { // '-'
      sign = -1;
      _ptr++;
    }
    int res = 0;
    while (_ptr < _bytes.length) {
      int c = _bytes[_ptr];
      if (c < 48 || c > 57) break;
      res = res * 10 + (c - 48);
      _ptr++;
    }
    return res * sign;
  }

  double readDouble() {
    while (_ptr < _bytes.length && _bytes[_ptr] <= 32) {
      _ptr++;
    }
    int start = _ptr;
    while (_ptr < _bytes.length && _bytes[_ptr] > 32) {
      _ptr++;
    }
    return double.parse(String.fromCharCodes(_bytes, start, _ptr));
  }
}

void solve(Scanner sc) {
  final buffer = StringBuffer();
  if (!sc.hasNext()) return;
  int t = sc.readInt();

  final solver = Solver();
  for (int i = 0; i < t; i++) {
    solver.solveTestCase(sc, buffer);
  }
  stdout.write(buffer.toString());
}

class Solver {
  // Flat arrays for points. Max N=10 -> 20 doubles.
  final Float64List points = Float64List(20);
  int n = 0;

  final Float64List compPoints = Float64List(20);
  int nComp = 0;

  final Float64List staticCands = Float64List(1000); // Max 500 candidates
  int nStaticCands = 0;

  final List<List<int>> staticCandsIndices = List.generate(10, (_) => []);

  final Float64List solution = Float64List(20);
  int nSol = 0;

  final Random _rng = Random(42);

  void solveTestCase(Scanner sc, StringBuffer buffer) {
    if (!sc.hasNext()) return;
    n = sc.readInt();

    for (int i = 0; i < n; i++) {
      points[2 * i] = sc.readDouble();
      points[2 * i + 1] = sc.readDouble();
    }

    int visited = 0;
    List<double> totalCircles = []; 

    for (int i = 0; i < n; i++) {
      if ((visited & (1 << i)) != 0) continue;

      nComp = 0;
      List<int> q = []; 
      q.add(i);
      visited |= (1 << i);

      int head = 0;
      while (head < q.length) {
        int u = q[head++];
        compPoints[2 * nComp] = points[2 * u];
        compPoints[2 * nComp + 1] = points[2 * u + 1];
        nComp++;

        for (int v = 0; v < n; v++) {
          if ((visited & (1 << v)) == 0) {
            double dx = points[2 * u] - points[2 * v];
            double dy = points[2 * u + 1] - points[2 * v + 1];
            if (dx * dx + dy * dy <= 16.0 + 1e-7) {
              visited |= (1 << v);
              q.add(v);
            }
          }
        }
      }

      if (!solveComponent(totalCircles)) {
        buffer.writeln("NO");
        return;
      }
    }

    buffer.writeln("YES");
    buffer.writeln(totalCircles.length ~/ 2);
    for (int i = 0; i < totalCircles.length; i += 2) {
      buffer.writeln("${totalCircles[i].toStringAsFixed(15)} ${totalCircles[i+1].toStringAsFixed(15)}");
    }
  }

  bool solveComponent(List<double> output) {
    // 1. K=1 Exact (MEC)
    if (getMEC(output)) return true;

    // 2. Backtracking (Deterministic)
    if (runBacktrack(false, output)) return true;

    // 3. Random fallback
    for (int i = 0; i < 3; i++) {
      if (runBacktrack(true, output)) return true;
    }

    return false;
  }

  bool runBacktrack(bool useRandom, List<double> output) {
    nStaticCands = 0;

    // Type 1: Points
    for (int i = 0; i < nComp; i++) {
      staticCands[2 * nStaticCands] = compPoints[2 * i];
      staticCands[2 * nStaticCands + 1] = compPoints[2 * i + 1];
      nStaticCands++;
    }

    // Type 2: Intersections
    for (int i = 0; i < nComp; i++) {
      for (int j = i + 1; j < nComp; j++) {
        addIntersections(
          compPoints[2 * i], compPoints[2 * i + 1], 1.0,
          compPoints[2 * j], compPoints[2 * j + 1], 1.0
        );
      }
    }

    if (useRandom) {
      for (int i = 0; i < nComp; i++) {
        for (int k = 0; k < 5; k++) {
          double angle = _rng.nextDouble() * 2 * pi;
          double r = sqrt(_rng.nextDouble()); 
          double cx = compPoints[2 * i] + r * cos(angle);
          double cy = compPoints[2 * i + 1] + r * sin(angle);
          staticCands[2 * nStaticCands] = cx;
          staticCands[2 * nStaticCands + 1] = cy;
          nStaticCands++;
        }
      }
    }

    // Pre-calculate indices
    for (int i = 0; i < nComp; i++) {
      staticCandsIndices[i].clear();
    }
    for (int idx = 0; idx < nStaticCands; idx++) {
      double cx = staticCands[2 * idx];
      double cy = staticCands[2 * idx + 1];
      for (int i = 0; i < nComp; i++) {
        double dx = cx - compPoints[2 * i];
        double dy = cy - compPoints[2 * i + 1];
        if (dx * dx + dy * dy <= 1.0 + EPS) {
          staticCandsIndices[i].add(idx);
        }
      }
    }

    nSol = 0;
    if (backtrack(0)) {
      for (int i = 0; i < nSol; i++) {
        output.add(solution[2 * i]);
        output.add(solution[2 * i + 1]);
      }
      return true;
    }
    return false;
  }
  
  void addIntersections(double x1, double y1, double r1, double x2, double y2, double r2) {
    double dx = x1 - x2;
    double dy = y1 - y2;
    double d2 = dx * dx + dy * dy;
    double d = sqrt(d2);

    if (d > r1 + r2 + EPS || d < (r1 - r2).abs() - EPS || d < 1e-9) return;

    double a = (r1 * r1 - r2 * r2 + d2) / (2 * d);
    double h = sqrt(max(0.0, r1 * r1 - a * a));

    double x0 = x1 + a * (x2 - x1) / d;
    double y0 = y1 + a * (y2 - y1) / d;

    double rx = -h * (y2 - y1) / d;
    double ry = h * (x2 - x1) / d;

    staticCands[2 * nStaticCands] = x0 - rx;
    staticCands[2 * nStaticCands + 1] = y0 - ry;
    nStaticCands++;

    staticCands[2 * nStaticCands] = x0 + rx;
    staticCands[2 * nStaticCands + 1] = y0 + ry;
    nStaticCands++;
  }

  bool backtrack(int mask) {
    if (mask == (1 << nComp) - 1) return true;

    int u = -1;
    for (int i = 0; i < nComp; i++) {
      if ((mask & (1 << i)) == 0) {
        u = i;
        break;
      }
    }

    bool tryCand(double cx, double cy) {
      double dx = cx - compPoints[2 * u];
      double dy = cy - compPoints[2 * u + 1];
      if (dx * dx + dy * dy > 1.0 + EPS) return false;

      for (int i = 0; i < nSol; i++) {
        double sx = solution[2 * i];
        double sy = solution[2 * i + 1];
        double d2 = (cx - sx) * (cx - sx) + (cy - sy) * (cy - sy);
        if (d2 < 4.0 - EPS) return false;
      }

      int newMask = mask;
      for (int i = 0; i < nComp; i++) {
        double px = compPoints[2 * i];
        double py = compPoints[2 * i + 1];
        if ((cx - px) * (cx - px) + (cy - py) * (cy - py) <= 1.0 + EPS) {
          newMask |= (1 << i);
        }
      }

      solution[2 * nSol] = cx;
      solution[2 * nSol + 1] = cy;
      nSol++;
      if (backtrack(newMask)) return true;
      nSol--;
      return false;
    }

    // 1. Static Candidates
    List<int> indices = staticCandsIndices[u];
    for (int i = 0; i < indices.length; i++) {
      int idx = indices[i];
      if (tryCand(staticCands[2 * idx], staticCands[2 * idx + 1])) return true;
    }

    // 2. Dynamic Candidates
    if (nSol > 0) {
       for (int i = 0; i < nSol; i++) {
          double sx = solution[2 * i];
          double sy = solution[2 * i + 1];

          double dux = compPoints[2 * u] - sx;
          double duy = compPoints[2 * u + 1] - sy;
          if (dux * dux + duy * duy > 9.0 + 1e-5) continue;
          
          if (tryIntersections(compPoints[2 * u], compPoints[2 * u + 1], 1.0, sx, sy, 2.0, tryCand)) return true;

          for (int k = 0; k < nComp; k++) {
             if (k == u) continue;
             double dx = compPoints[2 * k] - compPoints[2 * u];
             double dy = compPoints[2 * k + 1] - compPoints[2 * u + 1];
             if (dx * dx + dy * dy > 4.0 + 1e-5) continue;
             
             if (tryIntersections(compPoints[2 * k], compPoints[2 * k + 1], 1.0, sx, sy, 2.0, tryCand)) return true;
          }
       }
       
       for (int i = 0; i < nSol; i++) {
          double sx1 = solution[2 * i];
          double sy1 = solution[2 * i + 1];
          double dux1 = compPoints[2 * u] - sx1;
          double duy1 = compPoints[2 * u + 1] - sy1;
          if (dux1 * dux1 + duy1 * duy1 > 9.0 + 1e-5) continue;

          for (int j = i + 1; j < nSol; j++) {
             double sx2 = solution[2 * j];
             double sy2 = solution[2 * j + 1];
             double dux2 = compPoints[2 * u] - sx2;
             double duy2 = compPoints[2 * u + 1] - sy2;
             if (dux2 * dux2 + duy2 * duy2 > 9.0 + 1e-5) continue;

             if (tryIntersections(sx1, sy1, 2.0, sx2, sy2, 2.0, tryCand)) return true;
          }
       }
    }

    return false;
  }

  bool tryIntersections(double x1, double y1, double r1, double x2, double y2, double r2, bool Function(double, double) consumer) {
    double dx = x1 - x2;
    double dy = y1 - y2;
    double d2 = dx * dx + dy * dy;
    double d = sqrt(d2);

    if (d > r1 + r2 + EPS || d < (r1 - r2).abs() - EPS || d < 1e-9) return false;

    double a = (r1 * r1 - r2 * r2 + d2) / (2 * d);
    double h = sqrt(max(0.0, r1 * r1 - a * a));

    double x0 = x1 + a * (x2 - x1) / d;
    double y0 = y1 + a * (y2 - y1) / d;

    double rx = -h * (y2 - y1) / d;
    double ry = h * (x2 - x1) / d;

    if (consumer(x0 - rx, y0 - ry)) return true;
    if (consumer(x0 + rx, y0 + ry)) return true;
    return false;
  }
  
  bool getMEC(List<double> output) {
      if (nComp == 0) return false;
      if (nComp == 1) {
          output.add(compPoints[0]);
          output.add(compPoints[1]);
          return true;
      }
      
      double bestR2 = 1.0 + EPS;
      double bestX = 0, bestY = 0;
      bool found = false;
      
      void check(double cx, double cy, double r2) {
          if (r2 > bestR2) return;
          for (int i = 0; i < nComp; i++) {
              double dx = cx - compPoints[2 * i];
              double dy = cy - compPoints[2 * i + 1];
              if (dx * dx + dy * dy > r2 + EPS) return;
          }
          bestR2 = r2;
          bestX = cx;
          bestY = cy;
          found = true;
      }
      
      // Pairwise
      for (int i = 0; i < nComp; i++) {
          for (int j = i + 1; j < nComp; j++) {
              double cx = (compPoints[2 * i] + compPoints[2 * j]) / 2.0;
              double cy = (compPoints[2 * i + 1] + compPoints[2 * j + 1]) / 2.0;
              double dx = cx - compPoints[2 * i];
              double dy = cy - compPoints[2 * i + 1];
              check(cx, cy, dx * dx + dy * dy);
          }
      }
      
      // Triplet
      for (int i = 0; i < nComp; i++) {
          for (int j = i + 1; j < nComp; j++) {
              for (int k = j + 1; k < nComp; k++) {
                  double x1 = compPoints[2 * i], y1 = compPoints[2 * i + 1];
                  double x2 = compPoints[2 * j], y2 = compPoints[2 * j + 1];
                  double x3 = compPoints[2 * k], y3 = compPoints[2 * k + 1];
                  
                  double D = 2 * (x1 * (y2 - y3) + x2 * (y3 - y1) + x3 * (y1 - y2));
                  if (D.abs() < 1e-9) continue;
                  
                  double s1 = x1 * x1 + y1 * y1;
                  double s2 = x2 * x2 + y2 * y2;
                  double s3 = x3 * x3 + y3 * y3;
                  
                  double ux = (s1 * (y2 - y3) + s2 * (y3 - y1) + s3 * (y1 - y2)) / D;
                  double uy = (s1 * (x2 - x3) + s2 * (x3 - x1) + s3 * (x1 - x2)) / D;
                  
                  double dx = ux - x1;
                  double dy = uy - y1;
                  check(ux, uy, dx * dx + dy * dy);
              }
          }
      }
      
      if (found) {
          output.add(bestX);
          output.add(bestY);
          return true;
      }
      return false;
  }
}

File: 21/main.rs
use std::env;
use std::f64::consts::PI;
use std::io::{self, BufRead, BufReader, BufWriter, Write};
use std::time::{Duration, Instant};

const EPS: f64 = 1e-13;

#[derive(Clone, Copy, Debug)]
struct Point {
    x: f64,
    y: f64,
}

impl Point {
    fn new(x: f64, y: f64) -> Self {
        Point { x, y }
    }
}

fn dist_sq(p1: Point, p2: Point) -> f64 {
    let dx = p1.x - p2.x;
    let dy = p1.y - p2.y;
    dx * dx + dy * dy
}

struct Rng {
    state: u64,
}

impl Rng {
    fn new(seed: u64) -> Self {
        Self { state: seed }
    }

    fn next_u64(&mut self) -> u64 {
        let mut x = self.state;
        x ^= x << 13;
        x ^= x >> 7;
        x ^= x << 17;
        self.state = x;
        x
    }

    fn next_f64(&mut self) -> f64 {
        // Generates f64 in [0, 1)
        (self.next_u64() as f64) / (u64::MAX as f64)
    }
}

fn main() {
    let check_limits_env = env::var("CHECK_LIMITS").is_ok();
    
    let start_time = Instant::now();
    let max_time = Duration::from_secs(1000); // Local limit

    let stdin = io::stdin();
    let stdout = io::stdout();
    let mut reader = BufReader::new(stdin.lock());
    let mut writer = BufWriter::new(stdout.lock());

    let mut buffer = String::new();
    
    // Read T
    if reader.read_line(&mut buffer).is_err() { return; }
    let t_str = buffer.trim();
    if t_str.is_empty() { return; }
    let t: i32 = t_str.parse().unwrap_or(0);
    buffer.clear();

    let mut solver = Solver::new();

    for _ in 0..t {
        solver.solve_test_case(&mut reader, &mut writer, &mut buffer);
    }

    writer.flush().unwrap();

    if check_limits_env {
        let elapsed = start_time.elapsed();
        eprintln!("✓ Время: {:?}", elapsed);
    }
}

struct Solver {
    points: Vec<Point>,
    comp_points: Vec<Point>,
    static_cands: Vec<Point>,
    static_cands_indices: Vec<Vec<usize>>,
    solution: Vec<Point>,
    rng: Rng,
}

impl Solver {
    fn new() -> Self {
        Self {
            points: Vec::with_capacity(10),
            comp_points: Vec::with_capacity(10),
            static_cands: Vec::with_capacity(200),
            static_cands_indices: vec![Vec::with_capacity(50); 10],
            solution: Vec::with_capacity(10),
            rng: Rng::new(42),
        }
    }

    fn solve_test_case<R: BufRead, W: Write>(&mut self, reader: &mut R, writer: &mut W, buffer: &mut String) {
        // Read N
        loop {
            buffer.clear();
            if reader.read_line(buffer).unwrap() == 0 { return; }
            if !buffer.trim().is_empty() { break; }
        }
        let n: usize = buffer.trim().parse().unwrap_or(0);
        
        self.points.clear();
        for _ in 0..n {
            buffer.clear();
            reader.read_line(buffer).unwrap();
            let parts: Vec<&str> = buffer.trim().split_whitespace().collect();
            if parts.len() >= 2 {
                let x: f64 = parts[0].parse().unwrap();
                let y: f64 = parts[1].parse().unwrap();
                self.points.push(Point::new(x, y));
            }
        }

        let mut visited = 0;
        let mut total_circles = Vec::new();

        for i in 0..n {
            if (visited & (1 << i)) != 0 { continue; }

            self.comp_points.clear();
            let mut q = Vec::new();
            q.push(i);
            visited |= 1 << i;

            // BFS for component
            let mut head = 0;
            while head < q.len() {
                let u = q[head];
                head += 1;
                self.comp_points.push(self.points[u]);
                
                for v in 0..n {
                    if (visited & (1 << v)) == 0 {
                        if dist_sq(self.points[u], self.points[v]) <= 16.0 + 1e-7 {
                            visited |= 1 << v;
                            q.push(v);
                        }
                    }
                }
            }

            let res = self.solve_component();
            match res {
                Some(circles) => total_circles.extend(circles),
                None => {
                    writeln!(writer, "NO").unwrap();
                    return;
                }
            }
        }

        writeln!(writer, "YES").unwrap();
        writeln!(writer, "{}", total_circles.len()).unwrap();
        for c in total_circles {
            writeln!(writer, "{:.15} {:.15}", c.x, c.y).unwrap();
        }
    }

    fn solve_component(&mut self) -> Option<Vec<Point>> {
        let n_comp = self.comp_points.len();
        
        // 1. Try K=1 Exact (MEC)
        if let Some(center) = get_mec(&self.comp_points) {
            return Some(vec![center]);
        }

        // 2. Backtracking with Deterministic Candidates
        if let Some(res) = self.run_backtrack(n_comp, false) {
            return Some(res);
        }

        // 3. Backtracking with Random Candidates
        for _ in 0..3 {
            if let Some(res) = self.run_backtrack(n_comp, true) {
                return Some(res);
            }
        }

        None
    }

    fn run_backtrack(&mut self, n_comp: usize, use_random: bool) -> Option<Vec<Point>> {
        self.static_cands.clear();
        
        // Type 1: Points
        for i in 0..n_comp {
            self.static_cands.push(self.comp_points[i]);
        }

        // Type 2: Intersections
        for i in 0..n_comp {
            for j in i + 1..n_comp {
                let (p1, p2, cnt) = get_intersections(self.comp_points[i], 1.0, self.comp_points[j], 1.0);
                if cnt > 0 {
                    self.static_cands.push(p1);
                    if cnt > 1 {
                        self.static_cands.push(p2);
                    }
                }
            }
        }

        // Random candidates
        if use_random {
            for i in 0..n_comp {
                for _ in 0..5 {
                    let angle = self.rng.next_f64() * 2.0 * PI;
                    let r = self.rng.next_f64().sqrt() * 1.0;
                    let cx = self.comp_points[i].x + r * angle.cos();
                    let cy = self.comp_points[i].y + r * angle.sin();
                    self.static_cands.push(Point::new(cx, cy));
                }
            }
        }

        // Pre-calculate indices
        for i in 0..n_comp {
            self.static_cands_indices[i].clear();
        }
        
        for (idx, c) in self.static_cands.iter().enumerate() {
            for i in 0..n_comp {
                if dist_sq(*c, self.comp_points[i]) <= 1.0 + EPS {
                    self.static_cands_indices[i].push(idx);
                }
            }
        }

        self.solution.clear();
        if self.backtrack(0, n_comp) {
            return Some(self.solution.clone());
        }
        None
    }

    fn backtrack(&mut self, mask: u32, n_comp: usize) -> bool {
        if mask == (1 << n_comp) - 1 {
            return true;
        }

        let mut u = 0;
        for i in 0..n_comp {
            if (mask & (1 << i)) == 0 {
                u = i;
                break;
            }
        }

        // Try candidate closure
        let mut try_cand = |c: Point, solver: &mut Solver| -> bool {
            if dist_sq(c, solver.comp_points[u]) > 1.0 + EPS {
                return false;
            }
            for sol_c in &solver.solution {
                if dist_sq(c, *sol_c) < 4.0 - EPS {
                    return false;
                }
            }

            let mut new_mask = mask;
            for i in 0..n_comp {
                if dist_sq(c, solver.comp_points[i]) <= 1.0 + EPS {
                    new_mask |= 1 << i;
                }
            }

            solver.solution.push(c);
            if solver.backtrack(new_mask, n_comp) {
                return true;
            }
            solver.solution.pop();
            false
        };

        // 1. Static Candidates
        let indices = self.static_cands_indices[u].clone(); // Clone to avoid borrow conflict
        for &idx in &indices {
             if try_cand(self.static_cands[idx], self) {
                 return true;
             }
        }

        // 2. Dynamic Candidates
        if !self.solution.is_empty() {
             let sol_len = self.solution.len();
             // We need to collect dynamic candidates first to avoid borrowing conflicts
             // or restructure code. Collecting is safer/easier.
             let mut dyn_cands = Vec::with_capacity(50);

             // Type 3: Intersection of Boundary(P_i, 1) and Boundary(Sol_j, 2)
             for j in 0..sol_len {
                 let sol_c = self.solution[j];
                 if dist_sq(self.comp_points[u], sol_c) > 9.0 + 1e-5 {
                     continue;
                 }
                 
                 // Intersections involving u
                 let (p1, p2, cnt) = get_intersections(self.comp_points[u], 1.0, sol_c, 2.0);
                 if cnt > 0 {
                     dyn_cands.push(p1);
                     if cnt > 1 { dyn_cands.push(p2); }
                 }

                 // Other points close to u
                 for i in 0..n_comp {
                     if i == u { continue; }
                     if dist_sq(self.comp_points[u], self.comp_points[i]) > 4.0 + 1e-5 {
                         continue;
                     }
                     let (p1, p2, cnt) = get_intersections(self.comp_points[i], 1.0, sol_c, 2.0);
                     if cnt > 0 {
                         dyn_cands.push(p1);
                         if cnt > 1 { dyn_cands.push(p2); }
                     }
                 }
             }

             // Type 4: Intersection of two solution circles
             for i in 0..sol_len {
                 if dist_sq(self.comp_points[u], self.solution[i]) > 9.0 + 1e-5 { continue; }
                 for j in i+1..sol_len {
                     if dist_sq(self.comp_points[u], self.solution[j]) > 9.0 + 1e-5 { continue; }
                     let (p1, p2, cnt) = get_intersections(self.solution[i], 2.0, self.solution[j], 2.0);
                     if cnt > 0 {
                         dyn_cands.push(p1);
                         if cnt > 1 { dyn_cands.push(p2); }
                     }
                 }
             }

             for c in dyn_cands {
                 if try_cand(c, self) {
                     return true;
                 }
             }
        }

        false
    }
}

fn get_intersections(p1: Point, r1: f64, p2: Point, r2: f64) -> (Point, Point, usize) {
    let d2 = dist_sq(p1, p2);
    let d = d2.sqrt();
    if d > r1 + r2 + EPS || d < (r1 - r2).abs() - EPS || d < 1e-9 {
        return (Point::new(0.0, 0.0), Point::new(0.0, 0.0), 0);
    }
    
    let a = (r1 * r1 - r2 * r2 + d2) / (2.0 * d);
    let h = (r1 * r1 - a * a).max(0.0).sqrt();
    
    let x2 = p1.x + a * (p2.x - p1.x) / d;
    let y2 = p1.y + a * (p2.y - p1.y) / d;
    
    let rx = -h * (p2.y - p1.y) / d;
    let ry = h * (p2.x - p1.x) / d;

    (
        Point::new(x2 - rx, y2 - ry),
        Point::new(x2 + rx, y2 + ry),
        2
    )
}

fn get_mec(points: &[Point]) -> Option<Point> {
    let n = points.len();
    if n == 0 { return None; }
    if n == 1 { return Some(points[0]); }

    let mut best_r2 = 1.0 + EPS;
    let mut best_center: Option<Point> = None;

    let check = |c: Point, r2: f64, best_r2: &mut f64, best_center: &mut Option<Point>| {
        if r2 > *best_r2 { return; }
        for p in points {
            if dist_sq(c, *p) > r2 + EPS { return; }
        }
        *best_r2 = r2;
        *best_center = Some(c);
    };

    // Case 1: Pairwise
    for i in 0..n {
        for j in i + 1..n {
            let mid = Point::new((points[i].x + points[j].x) / 2.0, (points[i].y + points[j].y) / 2.0);
            let r2 = dist_sq(mid, points[i]);
            check(mid, r2, &mut best_r2, &mut best_center);
        }
    }

    // Case 2: Triplet
    for i in 0..n {
        for j in i + 1..n {
            for k in j + 1..n {
                if let Some(c) = get_circumcenter(points[i], points[j], points[k]) {
                    let r2 = dist_sq(c, points[i]);
                    check(c, r2, &mut best_r2, &mut best_center);
                }
            }
        }
    }

    best_center
}

fn get_circumcenter(a: Point, b: Point, c: Point) -> Option<Point> {
    let d = 2.0 * (a.x * (b.y - c.y) + b.x * (c.y - a.y) + c.x * (a.y - b.y));
    if d.abs() < 1e-9 { return None; }
    
    let sa = a.x * a.x + a.y * a.y;
    let sb = b.x * b.x + b.y * b.y;
    let sc = c.x * c.x + c.y * c.y;

    let ux = (sa * (b.y - c.y) + sb * (c.y - a.y) + sc * (a.y - b.y)) / d;
    let uy = (sa * (c.x - b.x) + sb * (a.x - c.x) + sc * (b.x - a.x)) / d;
    
    Some(Point::new(ux, uy))
}

